src/data/nat/cast.lean

```diff
@@ -262,73 +262,88 @@ by induction n; simp *
+lemma ext_nat' [add_monoid_hom_class F ℕ A] (f g : F) (h : f 1 = g 1) : f = g :=
+fun_like.ext f g $ begin
+  apply nat.rec,
+  { simp only [nat.nat_zero_eq_zero, map_zero] },
+  simp [nat.succ_eq_add_one, h] {contextual := tt}
+end
+
+@[ext] lemma add_monoid_hom.ext_nat : ∀ {f g : ℕ →+ A}, ∀ h : f 1 = g 1, f = g := ext_nat'
 
-variables [has_one A] [add_monoid B] [has_one B]
+variable [has_one A]
 
-lemma eq_nat_cast (f : ℕ →+ A) (h1 : f 1 = 1) :
-  ∀ n : ℕ, f n = n :=
-congr_fun $ show f = nat.cast_add_monoid_hom A, from ext_nat (h1.trans nat.cast_one.symm)
+-- these versions are primed so that the `ring_hom_class` versions aren't
+lemma eq_nat_cast' [add_monoid_hom_class F ℕ A] (f : F) (h1 : f 1 = 1) :
+  ∀ n : ℕ, f n = n
+| 0     := by simp
+| (n+1) := by rw [map_add, h1, eq_nat_cast' n, nat.cast_add_one]
 
-lemma map_nat_cast (f : A →+ B) (h1 : f 1 = 1) (n : ℕ) : f n = n :=
-(f.comp (nat.cast_add_monoid_hom A)).eq_nat_cast (by simp [h1]) _
+lemma map_nat_cast' [add_monoid_hom_class F A B] (f : F) (h : f 1 = 1) : ∀ (n : ℕ), f n = n
+| 0     := by simp
+| (n+1) := by rw [nat.cast_add, map_add, nat.cast_add, map_nat_cast', nat.cast_one, h, nat.cast_one]
 
-end add_monoid_hom
+end add_monoid_hom_class
 
-namespace monoid_with_zero_hom
+section monoid_with_zero_hom_class
 
-variables {A : Type*} [monoid_with_zero A]
+variables {A F : Type*} [monoid_with_zero A]
 
 /-- If two `monoid_with_zero_hom`s agree on the positive naturals they are equal. -/
-@[ext] theorem ext_nat {f g : monoid_with_zero_hom ℕ A}
+theorem ext_nat'' [monoid_with_zero_hom_class F ℕ A] (f g : F)
   (h_pos : ∀ {n : ℕ}, 0 < n → f n = g n) : f = g :=
 begin
-  ext (_ | n),
-  { rw [f.map_zero, g.map_zero] },
-  { exact h_pos n.zero_lt_succ, },
+  apply fun_like.ext,
+  rintro (_|n),
+  { simp },
+  exact h_pos n.succ_pos
 end
 
-end monoid_with_zero_hom
+@[ext] theorem monoid_with_zero_hom.ext_nat :
+∀ {f g : monoid_with_zero_hom ℕ A}, (∀ {n : ℕ}, 0 < n → f n = g n) → f = g := ext_nat''
+
+end monoid_with_zero_hom_class
 
-namespace ring_hom
+section ring_hom_class
 
-variables {R : Type*} {S : Type*} [non_assoc_semiring R] [non_assoc_semiring S]
+variables {R S F : Type*} [non_assoc_semiring R] [non_assoc_semiring S]
 
-@[simp] lemma eq_nat_cast (f : ℕ →+* R) (n : ℕ) : f n = n :=
-f.to_add_monoid_hom.eq_nat_cast f.map_one n
+@[simp] lemma eq_nat_cast [ring_hom_class F ℕ R] (f : F) : ∀ n, f n = n :=
+eq_nat_cast' f $ map_one f
 
-@[simp] lemma map_nat_cast (f : R →+* S) (n : ℕ) :
-  f n = n :=
-(f.comp (nat.cast_ring_hom R)).eq_nat_cast n
+@[simp] lemma map_nat_cast [ring_hom_class F R S] (f : F) : ∀ n : ℕ, f (n : R) = n :=
+map_nat_cast' f $ map_one f
 
-lemma ext_nat (f g : ℕ →+* R) : f = g :=
-coe_add_monoid_hom_injective $ add_monoid_hom.ext_nat $ f.map_one.trans g.map_one.symm
+lemma ext_nat [ring_hom_class F ℕ R] (f g : F) : f = g :=
+begin
+  apply @ext_nat' R, -- why does this need to be like this? even `ext_nat f g` doesn't work
+  simp only [map_one]
+end
 
```

Comment:
Most of the other cases should be fixable analogously, by putting in `show` to remind Lean that it knows the correct type all along.

Reply:
I think it's definitely better to wait for that fix to be merged, so that we can see how big the elaboration changes are. I'm going to remove `awaiting_review` until then. Thanks so much for the core lean fix!

Reply:
Can this now be reverted to:
```suggestion
ext_nat' f g (by simp only [map_one])
```
If not, I'd still prefer
```suggestion
ext_nat' f g $ show f 1 = g 1, by simp only [map_one]
```